Reproducible Android builds

pedropombeiro opened the pull request:
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

fixes #8203

### Summary

[comment]: # (Summarise the problem and how the pull request solves it)
This PR enables Reproducible Android builds by running the Android build end-to-end in a sandboxed Nix environment (which minimizes the influence of the host machine configuration on the build). To do so, I've:
- wrapped yarn dependencies in Nix using [yarn2nix](https://github.com/moretea/yarn2nix);
- wrapped Maven dependencies using a custom Nix script, for both clojars and gradle dependencies. This is a script that could be productized into a separate repo, but I haven't done so for lack of time;
- forked the logging library so that it doesn't cause the build path to leak into the JavaScript;
- patched the Go compiler in Nix so that the temp build path that it adds to the executable is controlled by Nix using an environment variable;

<!-- (Optional, remove if no changes to documentation) -->
Documentation change PR (review please): https://github.com/status-im/status.im/pull/375

### Review notes
<!-- (Optional. Specify if something in particular should be looked at, or ignored, during review) -->

The bulk of the work is done in the commit named `nix: Wrap gradle, npm packages, lein prod-build and gradle assembleRelease in Nix expressions to improve reproducible builds with constant build paths`

Notable changes with this PR:

- `make release-android` no longer deploys to connected Android device, it simply generates an .apk (much like `make release-ios` or `make release-desktop`). To deploy to the phone, you can call `make android-install`;
- for mobile dev environment, `node_modules` is now copied from the Nix store. A sentinel file called `node_modules/.copied~` is added to mark the latest time we modified the folder from a Nix script. If something in `node_modules/` changes after that, the directory is refreshed from the Nix store. Anytime `mobile_files/yarn.lock` changes, we need to run `make update-npm-nix` to regenerate the yarn2nix file;
- for Android release builds, the local Maven cache is not used. Instead, we use the artifacts downloaded into the Nix store (they can be recalculated using `make update-gradle-nix` and `make update-lein-nix`); 
- `scripts/prepare-for-platform.sh` is no longer needed for mobile platforms, that logic being moved to Nix scripts;
- `prod-build-android` and `release-android` are now built by Nix;
- added `disable-githooks` makefile target so that the CI server can disable them before the build;
- upgraded com.android.tools.build:gradle to 3.4.1;
- upgraded Gradle from 4.10 to 5.1.1;
- added several trimmed down Nix expressions to give us lighter-weight access to a specialized shell (e.g. for fastlane, watchman, adb, etc).

To avoid filing up the Nix cache, we should:
- not store the release builds nor the prod-builds (avoid uploading to Nix cache, and prune them regularly on CI servers);
- regularly run `nix-store --optimize` on CI servers and Nix cache server.

I've assigned this task to @jakubgs on [Wrike](https://www.wrike.com/open.htm?id=364672126).

### Testing notes
<!-- (Optional) -->

We need to do some general testing to see if the changes haven't broken anything, but normally nothing should have changed, except for the way we build the app.

In addition, we should verify that 2 android builds of the same commit, built on 2 different CI servers produce exactly the same apk file.

#### Platforms
<!-- (Optional. Specify which platforms should be tested) -->

- Android

<!-- (PRs will only be accepted if squashed into single commit.) -->

### Next steps

- Improve `make startdev-android-*` and other debug targets to use the Nix Maven repository, so that we have a single local cache and offline development is made possible;
status: ready <!-- Can be ready or wip -->

Commit:
nix: Make status-go mobile artifacts reproducible

- Replace random build paths in with fixed string
- Remove build ID sections

Commit:
nix: Wrap gradle, npm packages, `lein prod-build` and `gradle assembleRelease` in Nix expressions to improve reproducible builds with constant build paths

Commit:
Use forked com.taoensso/timbre for deterministic builds

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 0e4d5ac1 | [#1](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-8549/1/) | 2019-07-10 08:37:02 | ~4 min | `macos` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-8549/1/consoleText) |

jakubgs commented on Makefile:
```diff
@@ -41,90 +41,113 @@ endif
+_fix-perms: SHELL := /bin/sh
+_fix-perms: ##@prepare Fix permissions so that directory can be cleaned
+	$(shell test -d node_modules && chmod -R 744 node_modules)
+	$(shell test -d node_modules.tmp && chmod -R 744 node_modules.tmp)
```
Why inside `$(shell ...)`?

pedropombeiro replied:
because otherwise I get

```
test -d node_modules && chmod -R 744 node_modules
test -d node_modules.tmp && chmod -R 744 node_modules.tmp
Makefile:46: recipe for target '_fix-perms' failed
```

jakubgs commented on scripts/copy-translations.sh:
```diff
@@ -0,0 +1,23 @@
+
+cp -R "${GIT_ROOT}/translations" "${GIT_ROOT}/status-modules/"
+[ "$use_chmod" == 'chmod' ] && chmod u+w "${GIT_ROOT}/node_modules"
+cp -R "${GIT_ROOT}/status-modules" "${GIT_ROOT}/node_modules/"
```
Why are we copying them? Can't we use symlinks?

pedropombeiro replied:
It was like that in Roman's PR. I'll try with symlinks.

pedropombeiro replied:
Hmm, looks like `--link` isn't supported in macOS

jakubgs commented on Makefile:
```diff
@@ -41,90 +41,113 @@ endif
 release-ios: ##@build build release for iOS release
 	# Open XCode inside the Nix context
-	@$(MAKE) prod-build-ios && \
+	@watchman watch-del $${STATUS_REACT_HOME} && \
```
Wouldn't it make more sense to add `watchman-clean` target to dependency targets instead of running this by hand?

pedropombeiro replied:
Good point.

jakubgs commented on Makefile:
```diff
@@ -41,90 +41,113 @@ endif
-	lein prod-build-android && \
-	node prepare-modules.js
+	@git clean -dxf -f ./index.$(TARGET_OS).js && \
+	_NIX_RESULT_PATH=$(shell . ~/.nix-profile/etc/profile.d/nix.sh && nix-build --argstr target-os $(TARGET_OS) --pure --no-out-link --show-trace -A targets.mobile.prod-build) && \
```
Can you please add some comments here? Nobody sane will know what is happening in lines like this. Also, what is the git clean for?

pedropombeiro replied:
Done.
The git clean already existed for desktop target, if I recall correctly, so I just extended it to the rest of the targets. It's just to make sure that whatever we get is the result of the nix-build.

jakubgs commented on nix/shell.sh:
```diff
@@ -23,13 +23,23 @@ fi
+
+  entry_point='shell.nix'
+  attrset_option=''
+  if [ -n "${_NIX_ATTR}" ]; then
```
Please add a comment explaining what `_NIX_ATTR` is and what it does.

pedropombeiro replied:
Done.

jakubgs commented on Makefile:
```diff
@@ -41,90 +41,113 @@ endif
 	git clean -qdxf -f ./index.desktop.js desktop/ && \
 	lein prod-build-desktop && \
 	node prepare-modules.js
+	sed -i'' "s|$${STATUS_REACT_HOME}|.|g" index.desktop.js
```
Why? Comments please. Or put it in a separate Makefile target with a descriptive name and a comment and call it using `$(MAKE)`.

pedropombeiro replied:
This is to get rid of the build path from the JavaScript bundle. On second thought, this was premature optimization, we can worry this per-platform as we move each platform to being reproducible. I'll remove this from non-Android targets.

jakubgs commented on ci/Jenkinsfile.android:
```diff
@@ -52,24 +50,30 @@ pipeline {
     }
+    stage('Implicit dependencies') {
+      steps {
+        /* Build implicit dependencies if needed (we run `lein deps :tree` but it's not really required, for this purpose) */
```
I don't get what this comment means, at all.

pedropombeiro replied:
Expanded the comment with:
> Implicit dependencies include building a patched node_modules, fetching maven dependencies, and anything else required.
           We do this before the parallel steps so we have a known starting situation.

jakubgs commented on ci/Jenkinsfile.android:
```diff
@@ -102,14 +106,12 @@ pipeline {
   post {
     success { script { load('ci/github.groovy').notifyPR(true) } }
     failure { script { load('ci/github.groovy').notifyPR(false) } }
+    always { sh 'make _fix-perms' }
```
I thought you decided to drop read-only `node_modules`.

pedropombeiro replied:
I did, but I still need to copy the read-only version from Nix store into `node_modules.tmp`, which then gets moved into `node_modules`. If the build is stopped while that is happening, we'll end up with read-only `node_modules.tmp`, so better to have this to avoid having to fix things manually every now and then.

jakubgs commented on ci/Jenkinsfile.fastlane.clean:
```diff
@@ -23,7 +23,7 @@ pipeline {
       steps { script {
         nix = load('ci/nix.groovy')
-        nix.shell('bundle install --gemfile=fastlane/Gemfile')
+        nix.shell('bundle install --gemfile=fastlane/Gemfile', attr: 'targets.mobile.fastlane.shell')
```
Please break up long and multi-argument lines like this:
```groovy
nix.shell(
  'bundle install --gemfile=fastlane/Gemfile',
  attr: 'targets.mobile.fastlane.shell'
)
```

jakubgs commented on ci/Jenkinsfile.android:
```diff
@@ -32,9 +32,7 @@ pipeline {
     /* We use EXECUTOR_NUMBER to avoid multiple instances clashing */
-    LEIN_HOME         = "/var/tmp/lein-${EXECUTOR_NUMBER}"
-    YARN_CACHE_FOLDER = "/var/tmp/yarn-${EXECUTOR_NUMBER}"
-    GRADLE_USER_HOME  = "/var/tmp/gradle-${EXECUTOR_NUMBER}"
```
We are removing Yarn cache env variable because we use `yarn2nix` and not Yarn, correct? But why do we remove `GRADLE_USER_HOME`? Because Gradle no longer downloads anything?

pedropombeiro replied:
Correct on yarn. For Gradle, we just leave the default Gradle user home which means it'll end up in the Nix build directory. Does that make sense?

jakubgs replied:
Makes sense.

jakubgs commented on ci/Jenkinsfile.nix-cache:
```diff
@@ -35,9 +35,16 @@ pipeline {
-          nix-shell --pure --run echo
+          nix build -v --no-link --show-trace && \
+          nix-shell --pure --run echo --show-trace && \
+          nix-build --pure --no-out-link \$args
```
PLEASE add some comments, why are we doing this? Also, couldn't this be a separate stage? Why are you packing up the `Build` stage so much, it's always better to split those things so you know which part fails at a glance.

pedropombeiro replied:
Good point, thanks for catching it. This is a case where I did this while focused on getting stuff to work. With this week's radical change of focus, I'm kind of rushing to get this PR reviewed so it doesn't end up collecting dust. I'd normally have taken a bit more time going over things to make sure stuff like this hadn't fallen through the cracks.

pedropombeiro replied:
https://ci.status.im/job/status-react/job/nix-cache/job/macos/

jakubgs commented on ci/Jenkinsfile.nix-cache:
```diff
@@ -35,9 +35,16 @@ pipeline {
+          prodBuildSrcPath=\$(nix-store -q --binding src \$prodBuildDrv) && \
+          prodBuildOutPath=\$(nix-store -q --outputs \$prodBuildDrv) && \
+          nix-store --delete \$prodBuildDrv \$prodBuildSrcPath \$prodBuildOutPath
         """
```
Also, don't just put everything in one `sh` block, that's awful. Use a `script {}` Groovy block and call commands separately if you can.

jakubgs commented on ci/Jenkinsfile.nix-cache:
```diff
@@ -46,8 +53,14 @@ pipeline {
+              -and -not -name '*-status-react-release-android-source' \
+              -and -not -name '*-prod-build-*' \
+              -and -not -name '*-release-android' | \
+              xargs nix-copy-closure -v --to ${NIX_CACHE_USER}@${NIX_CACHE_HOST} && \
```
Wow, this is really getting out of control. I wonder if there's a better way of doing this.
Also, we are already doing `nix-store --delete` in the stage before, why then do we need those excludes here?

pedropombeiro replied:
For stuff that is already present in the machine (which won't get covered by a new delete). If we gc the build hosts, then I guess we don't need it.

jakubgs commented on ci/android.groovy:
```diff
@@ -31,18 +32,30 @@ def bundle() {
   }
-  sh 'find android/app/build/outputs/apk'
-  def outApk = "android/app/build/outputs/apk/${target}/app-${target}.apk"
+  sh 'find android/result/'
```
We can drop this, it's just some debug command.

jakubgs commented on ci/ios.groovy:
```diff
@@ -20,7 +20,8 @@ def bundle() {
     plutil('CFBundleVersion', utils.genBuildNumber()) +
-    plutil('CFBundleBuildUrl', currentBuild.absoluteUrl)
+    plutil('CFBundleBuildUrl', currentBuild.absoluteUrl),
+    attr: 'targets.mobile.ios.shell'
```
If we are doing the `attr` explicitly, we could just make each `plutil` call it's own `nix.shell` call, since it's not that expensive now.

jakubgs commented on ci/nix.groovy:
```diff
@@ -15,23 +15,107 @@ def shell(Map opts = [:], String cmd) {
+ *  - pure - Use --pure mode with Nix for more deterministic behaviour
+ *  - keep - List of env variables to pass through to Nix build
+ *  - args - Map of arguments to provide to --argstr
+ *  - attr - Name of attribute to use with --attr flag
```
Update the comments with new arguments and their meaning.

jakubgs commented on ci/nix.groovy:
```diff
@@ -15,23 +15,107 @@ def shell(Map opts = [:], String cmd) {
+  def keepFlags = []
+  def entryPoint = "\'${env.WORKSPACE}/shell.nix\'"
+  if (!isShell || opts.attr != null) {
+    entryPoint = "\'${env.WORKSPACE}/default.nix\'"
```
This does not need to be here at all, it can be added after the `nix-build` or `nix-shell` and the `_getNixCommandArgs` call in the parent method.

pedropombeiro replied:
I thought I had already explained why it makes sense to have it in the function that returned the args, but we can discuss it in a Meet.

jakubgs commented on ci/nix.groovy:
```diff
@@ -15,23 +15,107 @@ def shell(Map opts = [:], String cmd) {
+    def envVarsList = opts.keep.collect { var -> "${var}=\"${env[var]}\";" }
+    keepFlags = ["--arg env \'{${envVarsList.join("")}}\'"]
+
+    File envFile = File.createTempFile("nix-env", ".tmp")
```
Could we extract this whole thing related to creation of this tempfile into a separate method like `makeNixBuildEnvFile()` or something? This code is hard to read as it is.

jakubgs commented on ci/utils.groovy:
```diff
@@ -57,12 +57,20 @@ def doGitRebase() {
   return number
 }
 
+def getBuildNumber() {
```
Why is this method defined twice? Keep one or the other.

pedropombeiro replied:
It's not defined twice, one is called `genBuildNumber` (which creates the `BUILD_NUMBER` file) and the other is `getBuildNumber` (which reads it). Maybe I can rename it to `readBuildNumber` instead.

jakubgs replied:
Oh, I guess I'm blind, fair enough.

jakubgs commented on nix/derivation.nix:
```diff
@@ -0,0 +1,66 @@
+
+  # TARGETS
+  leiningen-shell = pkgs.mkShell (shellBootstraper {
+    buildInputs = with pkgs; [ clojure leiningen maven nodejs ];
```
Why are you passing `buildInputs` to `shellBootstraper` if it doesn't use it as far as I can tell? Does this syntax somehow passes it to `pkgs.mkShell`?

pedropombeiro replied:
It does.

jakubgs commented on nix/go/default.nix:
```diff
@@ -0,0 +1,20 @@
+		if tmp == "" {
+			tmp, err = ioutil.TempDir(os.Getenv("GOTMPDIR"), "go-build")
+		}'
+      # Disable chown tests, they fail with 'invalid argument'
```
You're mixing tabs and spaces my dude.

pedropombeiro replied:
I believe that's because the file we're replacing stuff in uses tabs. If we don't care I can just use spaces here.

jakubgs commented on nix/derivation.nix:
```diff
@@ -0,0 +1,66 @@
+  stdenv = pkgs.stdenvNoCC;
+  maven = pkgs.maven;
+  baseGo = pkgs.go_1_11;
+  go = pkgs.callPackage ./go { inherit baseGo; };
```
Maybe small comment explaining we are patching the Go compiler here?

pedropombeiro replied:
Maybe better if the comment is inside go/default.nix so we don't duplicate it?

jakubgs replied:
No need to duplicate, just a short comment saying we patch the Go compiler, the details can be in `go/default.nix`.

pedropombeiro replied:
It's a smaller duplication, but still a duplication. If in the future we no longer patch the Go compiler but simply build a specific version of it, we'll still need to update both the caller and callee. At this point, I'd probably prefer to call the folder `./patched-go`, because that enforces us to change both places.

jakubgs replied:
That makes sense!

jakubgs commented on nix/derivation.nix:
```diff
@@ -0,0 +1,66 @@
+  desktop = pkgs.callPackage ./desktop { inherit target-os stdenv status-go pkgs nodejs; inherit (pkgs) darwin; go = baseGo; };
+  mobile = pkgs.callPackage ./mobile { inherit target-os config stdenv pkgs nodejs yarn status-go maven localMavenRepoBuilder mkFilter prod-build-fn; inherit (pkgs.xcodeenv) composeXcodeWrapper; };
+  status-go = pkgs.callPackage ./status-go { inherit target-os go buildGoPackage; inherit (mobile.ios) xcodeWrapper; androidPkgs = mobile.android.androidComposition; };
+  mkFilter = import ./tools/mkFilter.nix { inherit (stdenv) lib; };
```
`mkFilter` doesn't mean much by itself, a comment stating that this is a `Filter for repo files when copying to build dir` or something like that.

jakubgs commented on nix/derivation.nix:
```diff
@@ -0,0 +1,66 @@
+  status-go = pkgs.callPackage ./status-go { inherit target-os go buildGoPackage; inherit (mobile.ios) xcodeWrapper; androidPkgs = mobile.android.androidComposition; };
+  mkFilter = import ./tools/mkFilter.nix { inherit (stdenv) lib; };
+  localMavenRepoBuilder = pkgs.callPackage ./tools/maven/maven-repo-builder.nix { inherit (pkgs) stdenv; };
+  prod-build-fn = pkgs.callPackage ./targets/prod-build.nix { inherit stdenv pkgs target-os nodejs localMavenRepoBuilder mkFilter; };
```
I have no idea what this does. Please explain. I can't even find an explanation for what `targets/prod-build.nix` does inside the file.

pedropombeiro replied:
Adding a comment
```
# Import a function that can build a prod-build target with specified node dependencies Nix expression
```

jakubgs commented on nix/derivation.nix:
```diff
@@ -0,0 +1,66 @@
+      if target-os == "android" then mobile.android.shellHook else
+      if target-os == "ios" then mobile.ios.shellHook else "";
+  });
+  watchman-shell = pkgs.mkShell {
```
Why does watchman need a shell? Does this just exist so we can call `make watchman-clean` without having to init a full Nix shell? Of so, please add a comment stating this.
Also, since this is not part of things necessary for building `status-react` as a whole, wouldn't it make sense to move stuff like that to something like `nix/helpers.nix` or something, to separate them from what is actually needed?

pedropombeiro replied:
Some stuff is possible to isolate to a completely separate file (although at that point there's no guarantee you're using the same version as in the full shell), others need some stuff that is parameterized here (think if you need a gradle shell), that's why I decided to leave things inside the tree structure and use attribute sets. Happy to discuss further in a call.

jakubgs replied:
Fair enough, but a comment about a purpose would be nice.

jakubgs commented on nix/mobile/android/default.nix:
```diff
@@ -0,0 +1,60 @@
+
+  mavenAndNpmDeps = callPackage ./maven-and-npm-deps { inherit stdenv stdenvNoCC gradle bash zlib androidEnvShellHook localMavenRepoBuilder mkFilter projectNodePackage status-go; };
+  gradle = pkgs.gradleGen.gradleGen rec {
+    name = "gradle-5.1.1";
```
I wonder if it wouldn't make sense to extract stuff like this(specific commits, urls, versions for a tool) into it's own `nix/mobile/android/gradle.nix` file. Kinda like `android-env.nix` is.

pedropombeiro replied:
Already done that.

jakubgs commented on nix/tools/mkFilter_test.nix:
```diff
@@ -0,0 +1,37 @@
+
+# This file serves as an example of the syntax required to call a Nix function, and serves to test mkFilter.nix
+# 
+# nix-shell --show-trace nix/tools/mkFilter_test.nix
```
I think my version was cooler:
https://gist.github.com/jakubgs/94d6d823b802691fb7a80afb29b450ac
Because it shows how to use `nix-instantiate` to test stuff without having to call `mkShell`.

pedropombeiro replied:
This file's your baby man, whatever you want.

jakubgs commented on nix/mobile/android/default.nix:
```diff
@@ -0,0 +1,60 @@
+  # TARGETS
+  release = callPackage ./targets/release-android.nix { inherit target-os gradle androidEnvShellHook mavenAndNpmDeps mkFilter nodejs prod-build status-go zlib; };
+  generate-maven-and-npm-deps-shell = callPackage ./maven-and-npm-deps/maven/shell.nix { inherit gradle maven androidEnvShellHook projectNodePackage status-go; };
+  adb-shell = pkgs.mkShell {
```
Same here, another helper we don't use in the actual build, but just in some very specific `Makefile` targets, maybe something like `helpers.nix` would be good to not mix this stuff.

jakubgs commented on nix/mobile/android/default.nix:
```diff
@@ -0,0 +1,60 @@
+    mavenAndNpmDeps.shellHook + ''
+    $STATUS_REACT_HOME/scripts/generate-keystore.sh
+
+    $STATUS_REACT_HOME/nix/mobile/reset-node_modules.sh "${mavenAndNpmDeps.buildInputs}/project" || exit
```
What does this do? Comment?

jakubgs commented on nix/mobile/android/maven-and-npm-deps/default.nix:
```diff
@@ -0,0 +1,207 @@
+  # fake build to pre-download deps into fixed-output derivation
+  deps = 
+    let
+      projectDir = "$NIX_BUILD_TOP/project";
```
I assume `NIX_BUILD_TOP` is the root of the build directory nix creates? Can we clarify that with a comment?

pedropombeiro replied:
I was only going to comment stuff that is specific to our project, but I can do that, sure.

jakubgs commented on nix/mobile/android/maven-and-npm-deps/default.nix:
```diff
@@ -0,0 +1,207 @@
+            chmod -R u+w $d
+          done
+
+          # Copy fresh node_modules
```
Fresh how?

pedropombeiro replied:
Is `# Copy mint node_modules from Nix store` more explicit?

jakubgs commented on nix/mobile/android/maven-and-npm-deps/default.nix:
```diff
@@ -0,0 +1,207 @@
+          chmod -R u+w ${projectDir}
+
+          cp -R ${projectDir}/status-modules/ ${projectDir}/node_modules/status-modules/
+          cp -R ${projectDir}/translations/ ${projectDir}/node_modules/status-modules/translations/
```
Why not symlink?

pedropombeiro replied:
Gradle doesn't seem to like symlinks:

FAILURE: Build failed with an exception.
```
* What went wrong:
Execution failed for task ':app:bundleReleaseJsAndAssets'.
> Could not list contents of '/build/status-react-release-android-source/node_modules/status-modules/translations'. Couldn't follow symbolic link.
```

jakubgs commented on nix/mobile/android/maven-and-npm-deps/default.nix:
```diff
@@ -0,0 +1,207 @@
+          prevSet=$-
+          set -e
+
+          patchShebangs ${projectDir}
```
Comment PLZ.

jakubgs replied:
Oh, I see:
```
        # Patch the shebangs of the bundled modules to prevent them from
        # calling executables outside the Nix store as much as possible
```

jakubgs commented on nix/mobile/android/maven-and-npm-deps/default.nix:
```diff
@@ -0,0 +1,207 @@
+          export REACT_NATIVE_DEPENDENCIES="${reactNativeDepsDir}"
+          export STATUS_REACT_HOME="${projectDir}"
+
+          pushd ${projectDir}/android
```
Please explain.

pedropombeiro replied:
Explain what, exactly?

jakubgs commented on nix/mobile/android/maven-and-npm-deps/default.nix:
```diff
@@ -0,0 +1,207 @@
+
+    export STATUSREACT_NIX_MAVEN_REPO="${deps}/.m2/repository"
+  '';
+}
```
After reading this file I feel like we've made things far too complicated for the benefit that we get. I do not look forward to fixing any of this when it breaks.

pedropombeiro replied:
I personally think the risk of this breaking is relatively low, since the way we're interacting with platforms involved (Maven, Gradle) doesn't change that often. The bigger risk is that react-native/ReactAndroid changes its gradle files (I don't expect radical changes in this type of code either), but that'll be part of any future RN upgrade process.

pedropombeiro replied:
I'm not sure what you mean by "for the benefit that we get". Do you mean reproducible builds as a whole?

jakubgs commented on nix/mobile/android/maven-and-npm-deps/maven/fetch-maven-deps.sh:
```diff
@@ -0,0 +1,191 @@
+
+set -Eeu
+
+# This script takes care of generating/updating the maven-inputs.txt file.
```
I'm not even going to try to review this monster. I assume it works and I don't want to touch it EVER.
If it does start causing trouble, I think a better approach would be to `strace` the download process and instead of hardcoding possible sources of packages just find them in the calls that do the requests(dns resolution in this case I guess).

pedropombeiro replied:
I'm willing to timebox giving `strace` a shot, now that all the other problems are sorted out.

jakubgs commented on nix/mobile/android/maven-and-npm-deps/maven/shell.nix:
```diff
@@ -0,0 +1,26 @@
+      rm -rf ./node_modules
+    fi
+
+    cp -aL --copy-contents ${projectNodePackage}/libexec/${projectNodePackage.passthru.pname}/node_modules/. ./node_modules/
```
Why not symlink since this is an `mkShell` not `mkDerivation`.

jakubgs commented on nix/mobile/android/targets/release-android.nix:
```diff
@@ -0,0 +1,129 @@
+    # Copy index.*.js input file
+    cp -a --no-preserve=ownership ${prod-build}/index*.js $sourceRoot/
+
+    # Copy android/ directory
```
I thought `builtins.path` and `mkFIlter` was already taking care of this stuff.

pedropombeiro replied:
Well here we're copying from the prod-build-android step (or something like that, i'm on mobile now), not from the user repo.

jakubgs commented on nix/mobile/android/targets/release-android.nix:
```diff
@@ -0,0 +1,129 @@
+    substituteInPlace android/gradlew \
+      --replace 'exec gradle' "exec gradle -S -Dmaven.repo.local='${localMavenRepo}' --offline ${gradle-opts}"
+    substituteInPlace android/app/build.gradle \
+      --replace 'versionCode 9999' 'versionCode ${build-number}'
```
Oh, so we set it to `9999` earlier to avoid re-building stuff, but now in the release build we change it to the actual value. Messy, but I get it. Maybe a better solution would be to implement something in `build.gradle` to cover this, rather than do this kind of patching?

pedropombeiro replied:
Added a property to gradle.properties:
```
versionCode=9999
```

This property gets replaced in a release-android build.

jakubgs commented on nix/mobile/android/targets/release-android.nix:
```diff
@@ -0,0 +1,129 @@
+          lib.mapAttrs (name: value:
+            "export ${name}='${value}';"
+            ) env));
+      unsetEnvVars = lib.concatStrings (
```
Can you explain these `lib.concatStrings` here with comment? I'm at a loss on what they do.

pedropombeiro replied:
I'm simplifying this code as follows:

```nix
      # Take the env attribute set and build a couple of scripts
      #  (one to export the environment variables, and another to unset them)
      exportEnvVars = concatStringsSep ";" (mapAttrsToList (name: value: "export ${name}='${value}'") env);
      unsetEnvVars = concatStringsSep ";" (mapAttrsToList (name: value: "unset ${name}") env);
```

jakubgs commented on shell.nix:
```diff
@@ -3,20 +3,15 @@
 let
-  projectDeps = import ./default.nix { inherit target-os pkgs nixpkgs-bootstrap; inherit (nixpkgs-bootstrap) config; };
+  project = import ./default.nix { inherit target-os pkgs nixpkgs-bootstrap; inherit (nixpkgs-bootstrap) config; };
+  projectShell = project.shell;
```
This variable seems pointless.

pedropombeiro replied:
Indeed, removing.

jakubgs commented on shell.nix:
```diff
@@ -3,20 +3,15 @@
-  };
 
-in mkShell {
+in mkShell (shellBootstraper {
```
So `mkShell` inherits here things like `buildInputs` passed to `shellBootstrapper` or am I missing something?

pedropombeiro replied:
Yeah, `shellBootstraper` modifies an attribute set and passes that to `mkShell`.

jakubgs commented on nix/tools/lein/generate-nix.sh:
```diff
@@ -0,0 +1,23 @@
+# For this, we start with a clean cache (in ./.m2~/repository/) and call cljsbuild
+#  to cause it to download all the artifacts. At the same time, we note them
+#  in lein-project-deps-maven-inputs.txt so that we can use that as an input
+#  to maven-inputs2nix.sh
```
I REALLY think one thing we need to do is extract the maven stuff into it's own repo, because this is really hard to follow. It needs to have clearly defined inputs and outputs, because I can't fully follow this.

jakubgs commented on nix/mobile/default.nix:
```diff
@@ -9,19 +9,52 @@ let
+  androidPlatform = callPackage ./android { inherit config pkgs target-os mkFilter nodejs maven localMavenRepoBuilder projectNodePackage prod-build-fn; status-go = status-go.android; };
+  iosPlatform = callPackage ./ios { inherit config pkgs mkFilter xcodeWrapper projectNodePackage; status-go = status-go.ios; };
+  fastlane = callPackage ./fastlane { inherit stdenv target-os; };
+  selectedSources = [
```
Can you explain this variable to me?

pedropombeiro replied:
I was hoping the naming/usage would be sufficient :-) There are quite a few expressions from which we need to gather buildInputs and shellHooks. Initially we were doing that manually, but it made more sense to do something like `lib.catAttrs "buildInputs" selectedSources` to avoid code repetition between sources for buildInputs and shellHooks.

jakubgs replied:
Okay.

jakubgs commented on nix/mobile/default.nix:
```diff
@@ -9,19 +9,52 @@ let
+    lib.optional platform.targetAndroid androidPlatform ++
+    lib.optional platform.targetIOS iosPlatform;
+
+  yarn2nix = import (fetchFromGitHub {
```
I really think stuff like that should be its own separate file. It makes it much harder to read what is going on here.

pedropombeiro replied:
Agree.

jakubgs commented on nix/mobile/default.nix:
```diff
@@ -9,19 +9,52 @@ let
+    sha256 = "0h2kzdfiw43rbiiffpqq9lkhvdv8mgzz2w29pzrxgv8d39x67vr9";
+    name = "yarn2nix-source";
+  }) { inherit pkgs nodejs yarn; };
+  projectNodePackage = yarn2nix.mkYarnPackage {
```
Same here, I get that this invokes `yarn2nix` on our node dependencies, but all the given parameters seem to be hardcoded aside from `projectNodePackage`. This should be it's own file with clear comments on what it does.

pedropombeiro replied:
Yeah, I'll move yarn2nix and this expression to its own file.

jakubgs reviewed (changes requested):
Nix expressions are extremely hard to read if you don't know what is the purpose of specific attributes. It is not a procedural flow that can be read from top to bottom with a logical order of steps, unlike most languages. It is a bunch of attributes(variables?) defined and used in various places.

The structure of Nix expressions is much more like the structure of an endlessly branching out tree, rather than a text with quotes and annotations referencing other texts. Without proper comments for things that are not obvious it's really tough to figure out what is for what purpose. It seems to me like most lines that do something less than trivial require a comment like:
```nix
# A is modified because of X issue and used in C and B
someDamnAttribute = callPackage ./moar/crazy/stuff { inherit all the shit }
```
Otherwise this is just madness.

I'm not finished with this PR, I'm going to re-read it tomorrow.

Commit:
Fix yarn.lock

jakubgs commented on Makefile:
```diff
@@ -64,58 +61,82 @@ else
-	scripts/build-desktop.sh
+	scripts/copy-translations.sh && \
+	scripts/build-desktop.sh; \
+	watchman watch-del $${STATUS_REACT_HOME}
```
You can just do `$(MAKE) watchman-clean`.

jakubgs commented on nix/mobile/ios/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+let
+  src =
+    let path = ./../../..;
```
If it works it works, but I'm always kinda skittish around hardcoded relative paths, but I assume that we'd have to pass some kind of `repoRoot` variable around to avoid this. Not sure if its worth it, thoughts?

pedropombeiro replied:
Unfortunately, I'm not aware of an easy way to pass around a path like that, or to find it out. `locateDominatingFile` might be what we need, but its API is cumbersome and there are no usages in nixpkgs to draw inspiration from (maybe tellingly).

PombeirP commented:
Refactored the mkShell code a bit so that we embed the shell-bootstrapper logic in an overridden mkShell function and pass that everywhere instead of relying on `pkgs.mkShell`. That way we're sure that the behavior is consistent in any shell.

jakubgs commented on Makefile:
```diff
@@ -64,58 +61,82 @@ else
+
+update-npm-nix: SHELL := /bin/sh
+update-npm-nix: ##@nix Update node2nix expressions based on current package.json
+	nix/desktop/realm-node/generate-nix.sh
```
This is not gonna work with `SHELL := /bin/sh` without first doing `. ~/.nix-profile/etc/profile.d/nix.sh`:
```
 $ make update-npm-nix
nix/desktop/realm-node/generate-nix.sh
nix/desktop/realm-node/generate-nix.sh: line 41: node2nix: command not found
```
Same foes for other `update-*` targets.

pedropombeiro replied:
Good catch!

jakubgs commented on nix/shell.sh:
```diff
@@ -2,6 +2,12 @@
+#     of the Nix expressions.
+# - _NIX_ATTR: This attribute can be used to specify an attribute set
+#     from inside the expression in `default.nix`, allowing drilling down into a specific attribute.
+# - NIX_KEEP: This variable allows specifying which env vars to keep for Nix pure shell. 
```
Either call both `NIX_ATTR` and `NIX_KEEP` or `_NIX_ATTR` and `_NIX_KEEP`.
Truth be told `TARGET_OS` should probably be `NIX_TARGET_OS` too.

pedropombeiro replied:
I'm avoiding a possible name clash with nix's own environment variables, so I'll rename NIX_KEEP

jakubgs commented on nix/tools/mkFilter_test.nix:
```diff
@@ -0,0 +1,37 @@
+
+# This file serves as an example of the syntax required to call a Nix function, and serves to test mkFilter.nix
+# 
+# nix-instantiate --strict --json --eval nix/tools/mkFilter_test.nix
```
This is wrong now. Your version uses `mkShell` unnecessarily.
Look at https://gist.github.com/jakubgs/94d6d823b802691fb7a80afb29b450ac.
You can return anything from this file, you don't need `mkShell` at all.

pedropombeiro replied:
Replaced.

jakubgs commented on nix/targets/prod-build.nix:
```diff
@@ -0,0 +1,66 @@
+  mkFilter, clojure, leiningen, maven, nodejs, localMavenRepoBuilder }:
+
+# The Nix expression takes a second argument to specify the node dependencies
+{ projectNodePackage }:
```
Why is this a separate argument, and not part of the first?

pedropombeiro replied:
The first argument I pass when I import the expression, and the second I pass when I actually invoke it.

jakubgs replied:
Makes sense. Than maybe update the comment? It's not a second argument. It's an argument to function that is returned from the Nix expression that is the file.

jakubgs reviewed (approved):
It kinda annoys me how Nix expressions switch between `lowerCamelCase` and `list-stype-variables`.

flexsurfer commented on deps.edn:
```diff
@@ -1,14 +1,16 @@
 {:paths ["components/src" "src" "react-native/src/cljsjs" "resources"]
  :deps  {org.clojure/clojure         {:mvn/version "1.10.0"} ;; Keep in sync with version from Nix
          org.clojure/clojurescript   {:mvn/version "1.10.520"}
+         com.google.javascript/closure-compiler-unshaded {:mvn/version "v20190325"}
```
could you please elaborate on these two lines

pedropombeiro replied:
I added these lines to make it the same as in project.clj. If it doesn't make sense I can remove them (at the time I was considering using a plugin to allow us to use the edn file in project.clj).

pedropombeiro replied:
Actually, I'll just get rid of them, since they're not required in this PR. Thanks for pointing them out.

PombeirP commented:
> It kinda annoys me how Nix expressions switch between `lowerCamelCase` and `list-stype-variables`.

I've been thinking about that too. In terms of readability, I kinda prefer dash-separated, but in terms of editing, I think camel-cased works out better. But I think we'll need to start moving towards camel-cased, as I think that's what the community mostly uses.

Commit:
nix: Only include fastlane in main shell when targetting iOS.

This sidesteps an issue where Nix complains about mismatched hashes in `ruby2.5.3-unf_ext-0.0.7.6` expression for Android CI builds.

Signed-off-by: Pedro Pombeiro <pombeirp@users.noreply.github.com>

asemiankevich commented:
> In addition, we should verify that 2 android builds of the same commit, built on 2 different CI servers produce exactly the same apk file.

@jakubgs could you please assist here?

jakubgs commented:
@asemiankevich You can just build the same commit again by clicking "Build with Parameters" here:
https://ci.status.im/job/status-react/job/prs/job/android/job/PR-8549/

Also, this PR is too fucking big. I wish we could extract somethings out of it. Like implementing of `yarn2nix` in a separate PR.

PombeirP commented:
> Also, this PR is too fucking big. I wish we could extract somethings out of it. Like implementing of `yarn2nix` in a separate PR.

We probably could, but since we want to pivot from working on this ASAP, it's not likely that we'd get buy-in to do that.

asemiankevich commented:
looks good to me, also same commit built twice produces same apk, checked with 39 and 40 job https://ci.status.im/job/status-react/job/prs/job/android/job/PR-8549/
